Updates to Migrating C++/CLI projects to .NET Core and .NET 5+ (#37687)
dotnet/docs: docs/core/porting/cpp-cli.md

1. Replace `<CLRSupport>true</CLRSupport>` properties with `<CLRSupport>NetCore</CLRSupport>`. This property is often in configuration-specific property groups, so you might need to replace it in multiple places.
2. Replace `<TargetFrameworkVersion>` properties with `<TargetFramework>netcoreapp3.1</TargetFramework>`. Be sure to change the tag as well as the value.
3. Remove any .NET Framework references to `System`, `System.Data`, `System.Windows.Forms`, and `System.Xml`, like `<Reference Include="System" />`. .NET Core SDK assemblies are automatically referenced when using `<CLRSupport>NetCore</CLRSupport>`.
4. Update API usage in *.cpp* files, as necessary, to remove APIs unavailable to .NET Core. Because C++/CLI projects tend to be fairly thin interop layers, there are often not many changes needed. You can use the [.NET Portability Analyzer](../../standard/analyzers/portability-analyzer.md) to identify unsupported .NET APIs used by C++/CLI binaries just as with purely managed binaries.

### WPF and Windows Forms usage

> You can use an app with a native entry point that displays a Windows Forms form via C++/CLI. You can also migrate a solution with a managed entry point interoperating with native dependencies via C++/CLI.

Reviewer: I don't think I would add this since it's not entirely clear to me.

3. When linking, provide the .NET Core app host directory as a `LibPath`, so that *ijwhost.lib* can be found.
4. Copy *ijwhost.dll* from the .NET Core app host directory to the project's output directory.
5. Make sure a [runtimeconfig.json](https://github.com/dotnet/sdk/blob/main/documentation/specs/runtime-configuration-file.md) file exists for the first component of the application that will run managed code. For latest versions of Visual Studio, a *runtime.config* file is created and copied automatically.

   For older versions of Visual Studio, if the application has a native entry point, you need to manually create the following *runtimeconfig.json* file for the first C++/CLI library to use the .NET Core runtime. If a C++/CLI library is called from a managed entry point, the library doesn't need a *runtimeconfig.json* file, because the entry point assembly has one that's used when starting the runtime.

   ```json
   {
         "runtimeOptions": {
            "tfm": "netcoreapp3.1",
            "framework": {
               "name": "Microsoft.NETCore.App",
               "version": "3.1.0"
            }
         }
   }
   ```

>[!NOTE]

> >C++/CLI assemblies can be loaded multiple times, each time into a new <xref:System.Runtime.Loader.AssemblyLoadContext>. If the first time managed code in a C++/CLI assembly is executed is from a native caller, the assembly is loaded into a separate `AssemblyLoadContext`. If the first time managed code is executed is from a managed caller, the assembly is loaded into the same `AssemblyLoadContext` as the caller, usually the default.

Reviewer: ```suggestion
>C++/CLI assemblies that target .NET 7 or a later version are always loaded into the default `AssemblyLoadContext`. However, in .NET 6 and earlier versions, C++/CLI assemblies might be loaded multiple times, each time into a new <xref:System.Runtime.Loader.AssemblyLoadContext>. If the first time that managed code in a C++/CLI assembly is executed:
>
> - Is from a native caller, the assembly is loaded into a separate `AssemblyLoadContext`.
> - Is from a managed caller, the assembly is loaded into the same `AssemblyLoadContext` as the caller, usually the default.
```

>
>To always load your C++/CLI assembly into the default `AssemblyLoadContext`, you can add an "initialize" style call from your entry-point assembly to your C++/CLI assembly. For more information, see this [dotnet/runtime issue](https://github.com/dotnet/runtime/issues/61105).

> >

Reviewer: ```suggestion
```

> >This limitation is removed starting in .NET 7. C++/CLI assemblies that target .NET 7 or a later version are always loaded into the default `AssemblyLoadContext`.

Reviewer: ```suggestion
```
